### ISSUE TITLE
Improve static analysis of `require()` calls

### ISSUE THREAD

**Issue by tmcw** [jsdoc/jsdoc#863]
I'm not sure if this is a known issue, but it doesn't seem like jsdoc resolves any `require()` statements with node. For instance: https://gist.github.com/tmcw/bf5f616d1fcb08e69b98 - running `jsdoc index.js` on this code creates an empty documentation directory.

**Comment by hegemonic** [comment]
If I may restate your feature request: It sounds like you want JSDoc to automatically figure out that `index.js` should get the exact same documentation as `b.js`. Is that correct?

You may also want to look at the docs explaining [how to document Node.js modules](http://usejsdoc.org/howto-commonjs-modules.html).

**Comment by tmcw** [comment]
Yeah, I'm trying to decipher that... I've gotten as far as https://gist.github.com/193bc7ae0138bd0fa03d but it's still not quite exporting anything. The documentation for node.js modules is a little funky because all of the examples mention modules named things like `color/mixer`, which is not a valid name for something in npm.

**Comment by tmcw** [comment]
So my question is more like: _can jsdoc resolve calls to require(), or does it only parse files specifically given to it?_ If I have a [modular system that's built on require](https://github.com/Turfjs/turf/blob/master/index.js), can jsdoc find & collect annotations to build documentation?

**Comment by hegemonic** [comment]
> Yeah, I'm trying to decipher that... I've gotten as far as https://gist.github.com/193bc7ae0138bd0fa03d but it's still not quite exporting anything.

One issue is that you're using both a `@module` tag and an `@exports` tag. You would normally use one or the other, not both. In this case, you should remove the `@exports` tag.

> The documentation for node.js modules is a little funky because all of the examples mention modules named things like `color/mixer`, which is not a valid name for something in npm.

Yes, but if you have an npm module named `color` that contains the file `mixer/index.js`, you would indeed be able to `require('color/mixer');`. Think of that example as the documentation for `mixer/index.js`. (There's probably a way to make this point more clearly in the docs. I'll file a new issue to address that.)

> So my question is more like: can jsdoc resolve calls to require(), or does it only parse files specifically given to it?

The latter, but I don't think that prevents you from using JSDoc for your package. Let's work this out on the [jsdoc-users mailing list](https://groups.google.com/forum/#!forum/jsdoc-users) so other JSDoc users can contribute to the discussion. Please send a message to the list, and I'll try to respond promptly.

I'll keep this issue open as a feature request for better static analysis of `require()` calls.
